### PR TITLE
Splitting expenses

### PR DESCRIPTION
**Pull request by jyuw**
#### Description
Changes proposed in this pull request:

* adds algorithm for splitting expenses within a group 
* displays who owes whom 

#### What I have learned working on this feature: 
* pry is love, pry is life
* accessing hash values
* render vs redirect_to
* nested loops
* using FactoryBot with associations

#### Screenshots:

screen shot 2018-03-03 at 12 04 55 pm

### CODE REVIEW

File: app/views/groups/show.html.erb [hunk 1]
```diff
@@ -25,4 +25,6 @@
   <p> <%= expense.user.email %> paid <%= expense.amount %>kr for <%= expense.description %></p>
 <% end %>
 
-<h3> WHO OWES WHO!?!!?!</h3>
+<h3> Who owes whom? </h3>
+<% binding.pry %>
```
Comment (tochman) [inline app/views/groups/show.html.erb]: Please remove this ;-)

File: features/expenses_can_be_split.feature [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      | faraz@ca.com | Holgers trip |
+      | david@ca.com | Holgers trip |
+    And the following expenses have been added
+      | email        | description  | amount | group        |
+      | david@ca.com | motorcycles  | 2000   | Holgers trip |
+      | faraz@ca.com | Eiffel Tower | 600    | Holgers trip |
+      | jade@ca.com  | cheese      | 1000   | Holgers trip |
+    And I am logged in as "jade@ca.com"
+
+  Scenario:
```
Comment (tochman) [inline features/expenses_can_be_split.feature]: I would like to see a scenario title here.

File: features/expenses_can_be_split.feature [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    Given I am on the "Holgers trip" page
+    Then I should see "faraz@ca.com paid 600kr for Eiffel Tower"
+    And I should see "david@ca.com paid 2000kr for motorcycles"
+    And I should see "jade@ca.com paid 1000kr for cheese"
+    And I should see "faraz@ca.com owes david@ca.com 600"
+    And I should see "jade@ca.com owes david@ca.com 200"
+    And I should see "Total expenses per person: 1200kr"
+
+
+  Scenario:
```
Comment (tochman) [inline features/expenses_can_be_split.feature]: ..and here...

File: features/step_definitions/expenses_steps.rb [hunk 4]
```diff
@@ -4,11 +4,23 @@
...
   end
-end
\ No newline at end of file
+end
+
+Given("the following expenses have been added") do |table|
+  table.hashes.each do |group|
+    current_group = Group.find_by(name: group[:group])
+    user = User.find_by(email: group[:email])
+    expense = Expense.create(description: group[:description], amount: group[:amount])
```
Comment (tochman) [inline features/step_definitions/expenses_steps.rb]: Why not use the factory?

File: app/controllers/expenses_controller.rb [hunk 5]
```diff
@@ -17,4 +20,59 @@ def add_current_user_to_expense
...
+    all_group_expenses.each do |email, amount|
+      next if total_amounts_owed[email] == 0
+      if amount > @group_average
+        lenders << { email => total_amounts_owed[email] }
+      else
+        debtors << { email => total_amounts_owed[email] }
+      end
+    end
+
+    lenders.each do |lender|
```
Comment (diraulo) [inline app/controllers/expenses_controller.rb:53]: This block looks less cryptic from what I saw yesterday 🙂

File: app/views/groups/show.html.erb [hunk 6]
```diff
@@ -2,25 +2,39 @@
...
+<%= f.submit value: 'Add expense' %>
 <% end %>
 
 <% @group.expenses.each do |expense| %>
-  <p> <%= expense.user.email %> -> <%= expense.description %>: <%= expense.amount %>kr</p>
+  <p><%= expense.user.email %> paid <%= expense.amount %>kr for <%= expense.description %></p>
+<% end %>
+
+<% if @output %>
+  <h3>
```
Comment (diraulo) [inline app/views/groups/show.html.erb:31]: Why not keep this `h3` in a single line?

File: app/views/layouts/mailer.html.erb [hunk 7]
```diff
@@ -1,13 +0,0 @@
-<!DOCTYPE html>
```
Comment (diraulo) [inline app/views/layouts/mailer.html.erb:1]: Any reason why this file was removed? Does it have any effect on this feature? 🤔

**Review by diraulo — COMMENTED** [review]
I see you guys managed to break CI? 🤔 

Also, fix the merge conflict on this branch.

File: app/views/layouts/mailer.html.erb (hunk 7 above)
Comment (holgertidemand) [reply app/views/layouts/mailer.html.erb:1]: We didnit use it!  got it from devise i think!

Comment (diraulo) [reply app/views/layouts/mailer.html.erb:1]: This file comes from the rails scaffold..

Comment (holgertidemand) [reply app/views/layouts/mailer.html.erb:1]: still works with out it ;)

Comment (jyuw) [reply app/views/layouts/mailer.html.erb:1]: Oops :p should we put it back?

Comment (diraulo) [reply app/views/layouts/mailer.html.erb:1]: I guess it'll take a while before you learn not to touch things that aren't directly related to the feature you're working on huh? 🤔 🙂

Not a blocker really... Just trying to keep you guys in line

Comment (tochman) [reply app/views/layouts/mailer.html.erb]: I would put it back. Yeah, just put it back.

**Review by diraulo — APPROVED** [review]

**Pull request merged**